types/express-serve-static-core/index.d.ts

```diff
@@ -209,7 +209,7 @@ export type Errback = (err: Error) => void;
  *     app.get<ParamsArray>(/user\/(.*)/, (req, res) => res.send(req.params[0]));
  *     app.get<ParamsArray>('/user/*', (req, res) => res.send(req.params[0]));
  */
-export interface Request<P extends Params = ParamsDictionary, ResBody = any, ReqBody = any> extends http.IncomingMessage, Express.Request {
+export interface Request<P extends Params = ParamsDictionary, ResBody = any, ReqBody = any, ReqQuery = any> extends http.IncomingMessage, Express.Request {
```

Comment:
Can we default this to `unknown` or `PoorMansUnknown` (see my PR https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43427)?

Reply:
In fact this should default to the return type of `qs.parse`, since that's the default query parser. https://expressjs.com/en/api.html#app.settings.table

`qs.parse`'s return type is `any` 😢. How about:

```ts
type ParsedQs = { [key: string]: string | ParsedQs };
```

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43440/files

Reply:
As this is generic now, one could pass on the expected type. I think most of the times we know that query parameters to expect, and having it as `{ [key: string]: string | ParsedQs }` would would not allow more specific types. 

Also, have it as generic, would allow you to pass on `{ [key: string]: string | ParsedQs }` as the parameter type.

Reply:
Ah, I see, do you mean by default this can be `{ [key: string]: string | ParsedQs }` ? I think that makes sense.

Reply:
Ah, I see you mean to update the default value to `type ParsedQs = { [key: string]: string | ParsedQs };`. I think that should be a fine assumption to make.

Reply:
But that would require a dependency on `ParsedQs` Should that be the case?

Reply:
Also, would the query string return an array in case of `?a=foo&a=bar&a=zoo` ?

Reply:
> Also, would the query string return an array in case of `?a=foo&a=bar&a=zoo` ?

Good point, it should be:

```ts
type ParsedQs = { [key: string]: string | string[] | ParsedQs };
```

> But that would require a dependency on `ParsedQs` Should that be the case?

We don't need to depend on it—we can inline the type inside the Express module.

Reply:
I would prefer to do this update in a separate PR. This may cause multiple other libraries to be updated.

Reply:
Also, from the examples for `qs` it seems it can create array of objects as well 

```
var arraysOfObjects = qs.parse('a[][b]=c');
assert.deepEqual(arraysOfObjects, { a: [{ b: 'c' }] });
```

Reply:
> This may cause multiple other libraries to be updated.

It didn't in my PR, so it should be fine.

Reply:
Updating the default value to be anything other than `any` would cause the libraries to break and would require an update to code to use type refinement. I don't think we should update the default type. 

```
(req: Request , res: Response) => req.query.foo // Type: string | Query | string[] | Query[]
```

Now to use the `foo` query type I will have to update all the places, and use type refinement / type assertion to get rid of the type error. 

Probably should be done as part of a major release. I don't want to take that on for now.

Reply:
> Probably should be done as part of a major release. I don't want to take that on for now.

I understand your concern about breaking other people's projects or libraries. However, whilst I do like the idea of waiting for a new major version of the library before making breaking changes to the types (to reduce inconvenience), in my experience that just isn't practical.

DefinitelyTyped has no system in place to allow us to coordinate breaking changes so that they are all released together. How would we remember to make this change when updating the types to support the next major version of Express? I don't think there's any way to coordinate these things. If we forget to make the change then, we'll have to wait for another major version to come along, which could take a very long time.

Therefore, I really do believe it's a choice between leaving the types broken (because they use `any` which _completely disables_ type checking, thereby defeating the whole point of TypeScript) or fixing the types immediately.

[I've been here many times before](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41675#issuecomment-575835308)… 😒